src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -596,26 +604,26 @@ internal unsafe SocketError DoOperationReceiveMessageFrom(Socket socket, SafeSoc
                     pMessage->controlBuffer.Length = _controlBufferPinned.Length;
                 }
                 pMessage->flags = _socketFlags;
-            }
 
-            NativeOverlapped* overlapped = AllocateNativeOverlapped();
-            try
-            {
-                SocketError socketError = socket.WSARecvMsg(
-                    handle,
-                    Marshal.UnsafeAddrOfPinnedArrayElement(_wsaMessageBufferPinned, 0),
-                    out int bytesTransferred,
-                    overlapped,
-                    IntPtr.Zero);
+                NativeOverlapped* overlapped = AllocateNativeOverlapped();
+                try
+                {
+                    SocketError socketError = socket.WSARecvMsg(
+                        handle,
+                        Marshal.UnsafeAddrOfPinnedArrayElement(_wsaMessageBufferPinned, 0),
+                        out int bytesTransferred,
+                        overlapped,
+                        IntPtr.Zero);
 
-                return ProcessIOCPResultWithSingleBufferHandle(socketError, bytesTransferred, overlapped);
-            }
-            catch
-            {
-                _singleBufferHandleState = SingleBufferHandleState.None;
-                FreeNativeOverlapped(overlapped);
-                _singleBufferHandle.Dispose();
-                throw;
+                    return ProcessIOCPResultWithSingleBufferHandle(socketError, bytesTransferred, overlapped, cancellationToken);
```

Comment:
I realize this is existing code, but...

This seems to be the only place we call ProcessIOCPResultWithSingleBufferHandle when we *don't* actually have a single buffer. Is there any harm in that? If not, why don't we just call ProcessIOCPResultWithSingleBufferHandle everywhere?

cc @stephentoub

Reply:
There are multiple things in there:

> This seems to be the only place we call ProcessIOCPResultWithSingleBufferHandle when we don't actually have a single buffer.

1. I don't think the above is true, we call the method from `DoOperationReceiveSingleBuffer`, `DoOperationReceiveFromSingleBuffer`,  `DoOperationSendSingleBuffer`, `DoOperationSendToSingleBuffer`.

2. There was an actual issue with `DoOperationReceiveMessageFrom`: calling `ProcessIOCPResultWithSingleBufferHandle` is only necessary in the single buffer case, and since the buffer is no longer being pre-pinned before doing things that can throw, `_singleBufferHandle.Dispose()` was unnecessary in the catch block. Addressed this in 19f3efca5b862f17f7c7649821360f16e6a290a9

3. I think we should eventually merge `ProcessIOCPResult` and `ProcessIOCPResultWithSingleBufferHandle` into a single method. See https://github.com/dotnet/runtime/pull/46903#discussion_r560898027

Reply:
There was an actual issue with `DoOperationReceiveMessageFrom`: calling `ProcessIOCPResultWithSingleBufferHandle` is only necessary in the single buffer case, and since the buffer is no longer being pre-pinned before doing things that can throw, `_singleBufferHandle.Dispose()` was unnecessary in the catch block. Addressed this in 19f3efca5b862f17f7c7649821360f16e6a290a9

Reply:
Yes, that's the issue I was concerned about. Looks like you fixed it, thanks.